BackendServer/python/analysis.py

```diff
@@ -0,0 +1,163 @@
+
+    # Total points
+    output_df['Total_Auto_Points'] = output_df['Auto_Grid_Points'] + output_df['Auto_Station_Points'] + output_df['Mobility_Points']
+    output_df['Total_Tele_Points'] = output_df['Tele_Grid_Points'] + output_df['Tele_Station_Points'] + output_df['Assumed_Link_Points']
+    output_df['Total_Points'] = output_df['Total_Auto_Points'] + output_df['Total_Tele_Points']
+    
+    return output_df
+
+def calculate_match_analysis(Team_Number, db_connection, *, appendTo = {}):
+    # Read data from MySQL database
+    matches_df = pd.read_sql('SELECT * FROM matchData WHERE Team_Number = %s AND No_Show_Robot = FALSE', db_connection, params=(Team_Number,))
+
+    # Calculate additional scores
+    calculate_match_scores(matches_df, mutate=True)
+    
+    # Create output variable
+    analysis_output = appendTo
+    
+    # Insert team number
+    analysis_output['Team_Number'] = Team_Number
+
+    # Calculate scoring aggregates
+    for type_phase, prefixes in (
+        ('Auto',        ('Auto_Cone', 'Auto_Cube')),
+        ('Tele_Pieces', ('Tele_Cone', 'Tele_Cube')),
+        ('Tele_Cone',   ('Tele_Cone',)),
+        ('Tele_Cube',   ('Tele_Cube',))
+    ):
+        for level, suffixes in (
+            ('Total', ('Low', 'Mid', 'High')),
+            ('Low',   ('Low',)),
+            ('Mid',   ('Mid',)),
+            ('High',  ('High',))
+        ):
+            # Get names of all the columns we want
+            included_columns = [f"{prefix}_{suffix}" for prefix in prefixes for suffix in suffixes]
+
+            # Sum up all of the included columns, then find the mean or max of it
+            analysis_output[f'{type_phase}_{level}_Average'] = matches_df[included_columns].sum(axis=1).mean()
+            analysis_output[f'{type_phase}_{level}_Max'] =     matches_df[included_columns].sum(axis=1).max()
+
+    # Calculate station frequencies
+    tele_station_attempts = (matches_df['Tele_Station'].isin(TELE_ATTEMPTED_DOCK_STATES)).sum()
+    analysis_output['End_Balance_Frequency'] = div((matches_df['Tele_Station'] == TeleStation.engaged).sum(), tele_station_attempts)
+    analysis_output['End_Dock_Frequency'] = div((matches_df['Tele_Station'].isin(TELE_SUCCESSFUL_DOCK_STATES)).sum(), tele_station_attempts)
+
+    auto_station_attempts = (matches_df['Auto_Station'].isin(AUTO_ATTEMPTED_DOCK_STATES)).sum()
+    analysis_output['Auto_Balance_Frequency'] = div((matches_df['Auto_Station'] == AutoStation.engaged).sum(), auto_station_attempts)
+    analysis_output['Auto_Dock_Frequency'] = div((matches_df['Auto_Station'].isin(AUTO_SUCCESSFUL_DOCK_STATES)).sum(), auto_station_attempts)
+
+    # Calculate average period points
+    analysis_output['Average_Teleop_Points'] = matches_df['Total_Tele_Points'].mean()
+    analysis_output['Average_Auto_Points'] = matches_df['Total_Auto_Points'].mean()
+
+    # Convert to ints
+    for i in analysis_output:
```

Comment:
instead of iterating through the keys, you could directly iterate through the values with `for value in analysis_output.values():`

Reply:
that will save you the lookup each time - if you want key + value you could also use `analysis_output.items():`

Reply:
The problem is I have to alter the value in the dictionary (`analysis_output[i] = int(analysis_output[i])`) so I need at least one lookup